### ISSUE TITLE
Eval bug: --spec-type draft-mtp segfaults on Qwen3.6-35B-A3B (MoE) — works fine on dense Qwen3.8-27B

### ISSUE THREAD

**Issue by zhou51320** [mxxm-t/mx-llama.cpp#12]
### Name and Version

root@72f426cd317a:/# llama-server --version
version: 0.4.0-dev (build 10851, commit 67672dc5b)
built with GNU 13.3.0 for Linux x86_64

### Operating systems

Linux

### GGML backends

HIP

### Hardware

Hardware: 2x AMD MI50 (gfx906), 32GB each
ROCm 7.2.1 / mxxm/mx-llama.cpp:gfx906 (or built from latest master)

### Models

Qwen3.6-35B-A3B-UD-Q6_K.gguf

### Problem description & steps to reproduce

Command
llama-server --model Qwen3.6-35B-A3B-UD-Q6_K.gguf \
  --mmproj mmproj-F32.gguf --host 0.0.0.0 --port 9999 --api-key X \
  -ngl 99 -c 262144 -b 8192 -np 1 \
  --spec-type draft-mtp --spec-draft-n-max 3 \
  --cache-type-k q8_0 --cache-type-v q8_0 -lv 6

Reproduction
Server loads fine (model loaded, speculative context initialized).
First request only → Segmentation fault (core dumped), during the first decode right after prefill scheduling.

Test matrix
#	Config	Result
1	35B + draft-mtp (n-max 3)	Segfault
2	35B + draft-mtp (n-max 3) + GGML_CUDA_REPACK_KQUANT_MOE_FUSION=0	Segfault
3	35B + draft-mtp + --spec-draft-n-max 1	Segfault
4	35B, no speculative decoding	OK
5	Qwen3.8-27B (dense) + draft-mtp (n-max 3)	OK
Conclusion draft-mtp on a MoE target segfaults on the first decode; unrelated to the fused MoE repack (env-var escape hatch does not help). Dense models unaffected.

### Relevant log output

0.09.845.124 I clip_ctx: CLIP using ROCm0 backend
0.09.845.349 W load_hparams: Qwen-VL models require at minimum 1024 image tokens to function correctly on grounding tasks
0.09.845.349 W load_hparams: if you encounter problems with accuracy, try adding --image-min-tokens 1024
0.09.845.349 W load_hparams: more info: https://github.com/ggml-org/llama.cpp/issues/16842
0.09.845.349 I load_hparams: projector:          qwen3vl_merger
0.09.845.350 I load_hparams: n_embd:             1152
0.09.845.350 I load_hparams: n_head:             16
0.09.845.350 I load_hparams: n_ff:               4304
0.09.845.350 I load_hparams: n_layer:            27
0.09.845.350 I load_hparams: ffn_op:             gelu
0.09.845.350 I load_hparams: projection_dim:     2048
0.09.845.351 I
--- vision hparams ---
0.09.845.351 I load_hparams: image_size:         768
0.09.845.351 I load_hparams: patch_size:         16
0.09.845.351 I load_hparams: has_llava_proj:     0
0.09.845.351 I load_hparams: minicpmv_version:   0
0.09.845.351 I load_hparams: n_merge:            2
0.09.845.352 I load_hparams: n_wa_pattern: 0
0.09.845.352 I load_hparams: image_min_pixels:   8192
0.09.845.352 I load_hparams: image_max_pixels:   4194304
0.09.845.352 I
0.09.845.354 I load_hparams: model size:         1703.53 MiB
0.09.845.354 I load_hparams: metadata size:      0.12 MiB
0.10.371.127 D load_tensors: loaded 334 tensors from /app/models/Qwen3.6-35B-A3B-MTP/mmproj-F32.gguf
0.10.375.347 I get_dummy_batch: warmup with image size = 1472 x 1472
0.10.380.493 I get_dummy_batch: warmup with image size = 1472 x 1472
0.10.387.969 I reserve_compute_meta:      ROCm0 compute buffer size =   248.10 MiB
0.10.387.972 I reserve_compute_meta:        CPU compute buffer size =    24.93 MiB
0.10.387.973 I reserve_compute_meta: graph splits = 1, nodes = 823
0.10.387.992 I warmup: flash attention is enabled
0.10.388.002 I srv    load_model: loaded multimodal model, '/app/models/Qwen3.6-35B-A3B-MTP/mmproj-F32.gguf'
0.10.396.567 D CUDA Graph id 74 reused
0.10.396.569 D ggml_backend_cuda_graph_compute: CUDA graph warmup complete
0.10.415.980 D CUDA Graph id 75 reused
0.10.415.983 D ggml_backend_cuda_graph_compute: CUDA graph warmup complete
0.10.426.838 I cmn  common_conte: the context supports bounded partial sequence removal
0.10.440.561 I srv    load_model: initializing, n_slots = 1, n_ctx_slot = 262144, kv_unified = 'false'
0.10.440.579 I spec common_specu: adding speculative implementation 'draft-mtp'
0.10.440.582 I spec common_specu: - n_max=3, n_min=0, p_min=0.00, n_embd=2048, backend_sampling=1
0.10.440.582 I spec common_specu: - gpu_layers=-1, cache_k=f16, cache_v=f16, ctx_tgt=yes, ctx_dft=yes, devices=[default]
0.10.440.898 D set_sampler: seq_id = 0, sampler = 0x58eb12366dc0
0.10.440.933 D set_embeddings_nextn: value = 1, masked = 0
0.10.440.934 D set_embeddings_nextn: value = 1, masked = 1
0.10.443.755 I sched_reserve: reserving ...
0.10.443.759 D sched_reserve: max_nodes = 24101
0.10.449.348 D sched_reserve: reserving full memory module
0.10.449.354 D sched_reserve: worst-case: n_tokens = 512, n_seqs = 1, n_outputs = 1
0.10.449.356 D graph_reserve: reserving a graph for ubatch with n_tokens =  512, n_seqs =  1, n_outputs =    1
0.10.545.795 D graph_reserve: reserving a graph for ubatch with n_tokens =    1, n_seqs =  1, n_outputs =    1
0.10.546.028 D graph_reserve: reserving a graph for ubatch with n_tokens =  512, n_seqs =  1, n_outputs =    1
0.10.546.201 I sched_reserve:      ROCm1 compute buffer size =  1153.06 MiB
0.10.546.203 I sched_reserve:  ROCm_Host compute buffer size =  1044.07 MiB
0.10.546.203 I sched_reserve: graph nodes  = 92
0.10.546.204 I sched_reserve: graph splits = 2
0.10.546.204 I sched_reserve: reserve took 102.45 ms, sched copies = 4
0.10.559.778 I srv    load_model: speculative decoding context initialized
0.10.559.785 I slot   load_model: id  0 | task -1 | new slot, n_ctx = 262144
0.10.559.786 D slot        reset: id  0 | task -1 |
0.10.560.201 I srv    load_model: prompt cache is enabled, size limit: 8192 MiB
0.10.560.201 I srv    load_model: use `--cache-ram 0` to disable the prompt cache
0.10.560.202 I srv    load_model: for more info see https://github.com/ggml-org/llama.cpp/pull/16391
0.10.560.202 I srv    load_model: context checkpoints enabled, max = 32, min spacing = 8192
0.10.560.221 I srv          init: idle slots will be saved to prompt cache upon starting a new task
0.10.560.221 D srv          init: __TEST_TAG_CACHE_IDLE_SLOTS_ENABLED__
0.10.563.980 D Using specialized template: Qwen3-Coder
0.10.565.639 I srv          init: init: chat template, example_format: '<|im_start|>system
You are a helpful assistant<|im_end|>
<|im_start|>user
Hello<|im_end|>
<|im_start|>assistant
<think>
</think>
Hi there<|im_end|>
<|im_start|>user
How are you?<|im_end|>
<|im_start|>assistant
<think>
'
0.10.565.666 D Using specialized template: Qwen3-Coder
0.10.566.400 I srv          init: init: chat template, thinking = 1
0.10.566.410 I srv          init: preserve_reasoning kwarg: true
0.10.566.411 W srv          init: chat template supports preserving reasoning, it is enabled by default (may use more tokens, disable via --no-reasoning-preserve)
0.10.566.439 I srv  llama_server: model loaded
0.10.566.441 I srv  llama_server: listening on http://0.0.0.0:9999
0.10.566.497 D que    start_loop: processing new tasks
0.10.566.499 D que    start_loop: update slots
0.10.566.501 I srv  update_slots: all slots are idle
0.10.566.501 D que    start_loop: waiting for new tasks
0.12.614.302 D srv          stop: all tasks already finished, no need to cancel
0.12.691.677 D Using specialized template: Qwen3-Coder
0.12.696.947 I srv  server_strea: conv_id=ys1rbdfxwgn (empty=0)
0.12.696.977 D add_text: <|im_start|>system
# Tools
You have access to the following functions:
<tools>
{"type": "function", "function": {"name": "get_datetime", "description": "Returns the current local date and time in ISO 8601 format, with the IANA time zone name", "parameters": {"properties": {}, "required": [], "type": "object"}}}
{"type": "function", "function": {"name": "get_info", "description": "Get runtime info (OS name), may call when user asks about local files or shell commands", "parameters": {"properties": {}, "required": [], "type": "object"}}}
</tools>
If you choose to call a function ONLY reply in the following format with NO suffix:
<tool_call>
<function=example_function_name>
<parameter=example_parameter_1>
value_1
</parameter>
<parameter=example_parameter_2>
This is the value for the second parameter
that can span
multiple lines
</parameter>
</function>
</tool_call>
<IMPORTANT>
Reminder:
- Function calls MUST follow the specified format: an inner <function=...></function> block must be nested within <tool_call></tool_call> XML tags
- Required parameters MUST be specified
- You may provide optional reasoning for your function call in natural language BEFORE the function call, but NOT after
- If there is no function call available, answer the question like normal with your current knowledge and do not tell the user about function calls
</IMPORTANT><|im_end|>
<|im_start|>user
你的知识库最新是啥日期<|im_end|>
<|im_start|>assistant
<think>
0.12.698.039 D srv    operator(): Grammar (tool_calls): root ::= tool-call-root
space ::= | " " | "\n"{1,2} [ \t]{0,20}
tool-call ::= "<tool_call>\n" (tool-get-datetime | tool-get-info) "</tool_call>" space
tool-call-root ::= (tool-call tool-call*)?
tool-get-datetime ::= ("<function=" "get_datetime" ">\n") "</function>\n"
tool-get-info ::= ("<function=" "get_info" ">\n") "</function>\n"
0.12.698.040 I srv    operator(): chat format: peg-native
0.12.698.309 D srv  eval_llama_c: reasoning budget: tokens=-1, generation_prompt='<|im_start|>assistant
<think>
', start=1 toks, end=2 seqs, forced=1 toks
0.12.698.365 D res  add_waiting_: add task 0 to waiting list. current waiting = 0 (before add)
0.12.698.367 D que          post: new task, id = 0/1, front = 0
0.12.698.375 D que    start_loop: processing new tasks
0.12.698.382 D que  process_new_: processing task, id = 0
0.12.698.385 I slot get_availabl: id  0 | task -1 |  - skipping, slot is empty
0.12.698.386 I slot get_availabl: id  0 | task -1 | selected slot by LRU, t_last = -1
0.12.698.386 I srv  get_availabl: updating prompt cache
0.12.698.390 I srv          load:  - looking for better prompt, base f_keep = -1.000, f_sim = 0.000
0.12.698.393 I srv        update:  - cache state: 0 prompts, 0.000 MiB (limits: 8192.000 MiB, 262144 tokens, 8589934592 est)
0.12.698.394 I srv  get_availabl: prompt cache update took 0.01 ms
0.12.698.402 D slot launch_slot_: id  0 | task -1 | launching slot : {"id":0,"n_ctx":262144,"speculative":true,"is_processing":false}
0.12.698.478 D common_sampler_init: prefill token: 248045 = <|im_start|>
0.12.698.479 D common_sampler_init: prefill token: 74455 = assistant
0.12.698.480 D common_sampler_init: prefill token: 198 =
0.12.698.480 D common_sampler_init: prefill token: 248068 = <think>
0.12.698.480 D common_sampler_init: prefill token: 198 =
0.12.698.489 D common_sampler_init: reasoning-budget accepted prefill token (248045)
0.12.698.489 D common_sampler_init: reasoning-budget accepted prefill token (74455)
0.12.698.489 D common_sampler_init: reasoning-budget accepted prefill token (198)
0.12.698.489 I cmn  common_reaso: activated, budget=2147483647 tokens
0.12.698.490 D common_sampler_init: reasoning-budget accepted prefill token (248068)
0.12.698.490 D common_sampler_init: reasoning-budget accepted prefill token (198)
0.12.698.511 I slot launch_slot_: id  0 | task -1 | sampler chain: logits -> ?penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> top-p -> min-p -> ?xtc -> ?temp-ext -> dist
0.12.698.517 I slot launch_slot_: id  0 | task -1 | sampler params:
        repeat_last_n = 64, repeat_penalty = 1.000, frequency_penalty = 0.000, presence_penalty = 0.000
        dry_multiplier = 0.000, dry_base = 1.750, dry_allowed_length = 2, dry_penalty_last_n = 64
        top_k = 20, top_p = 0.950, min_p = 0.050, xtc_probability = 0.000, xtc_threshold = 0.100, typical_p = 1.000, top_n_sigma = -1.000, temp = 1.000
        mirostat = 0, mirostat_lr = 0.100, mirostat_ent = 5.000, adaptive_target = -1.000, adaptive_decay = 0.900
0.12.698.518 I slot launch_slot_: id  0 | task 0 | processing task, is_child = 0
0.12.698.519 D que    start_loop: update slots
0.12.698.519 D srv  update_slots: posting NEXT_RESPONSE
0.12.698.520 D que          post: new task, id = 1, front = 0
0.12.698.526 I slot   operator(): id  0 | task 0 | new prompt, n_ctx_slot = 262144, n_keep = 0, task.n_tokens = 338
0.12.698.530 D res          send: sending result for task id = 0
0.12.698.530 D res          send: task id = 0 pushed to result queue
0.12.698.533 I slot   operator(): id  0 | task 0 | cached n_tokens = 0, memory_seq_rm [0, end)
0.12.698.621 D srv          stop: all tasks already finished, no need to cancel
0.12.698.669 D srv    operator(): http: streamed chunk: data: {"choices":[{"finish_reason":null,"index":0,"delta":{"role":"assistant","content":null}}],"created":1788848458,"id":"chatcmpl-6VnEsgbBZeFiuak1yUneU6kQmcRlkrB2","model":"/app/models/Qwen3.6-35B-A3B-MTP/Qwen3.6-35B-A3B-UD-Q6_K.gguf","system_fingerprint":"b10851-67672dc5b","object":"chat.completion.chunk","timings":{"cache_n":0,"prompt_n":0,"prompt_ms":0.0,"prompt_per_token_ms":0.0,"prompt_per_second":0.0,"predicted_n":0,"predicted_ms":0.0,"predicted_per_token_ms":0.0,"predicted_per_second":0.0},"prompt_progress":{"total":338,"cache":0,"processed":0,"time_ms":0}}
0.12.698.839 D slot   operator(): id  0 | task 0 | main/do_checkpoint = no, pos_min = -1, pos_max = -1
0.12.698.849 D set_adapters_lora: adapters = (nil)
0.12.698.850 D adapters_lora_are_same: adapters = (nil)
0.12.698.851 D set_embeddings: value = 0
0.12.698.853 D srv        decode: n_batch (effective) = 8192, off = 0
0.12.698.853 D que  yield_to_que: yielding to queue
0.12.698.926 D que  process_new_: processing task, id = 1
0.12.698.928 D srv  process_sing: decoding, decline task, id_task = 1
0.12.698.928 D que  process_new_: task declined, id = 1
0.12.712.674 D ggml_backend_cuda_graph_compute: CUDA graph warmup reset
0.13.250.937 D ggml_backend_cuda_graph_compute: CUDA graph warmup reset
0.13.396.061 D que  yield_to_que: done yielding to queue
0.13.396.064 D que  yield_to_que: yielding to queue
0.13.396.118 D que  process_new_: processing task, id = 1
0.13.396.119 D srv  process_sing: decoding, decline task, id_task = 1
0.13.396.119 D que  process_new_: task declined, id = 1
Segmentation fault (core dumped)
root@72f426cd317a:/#

**Comment by zhou51320** [comment]
Sorry, I pulled the wrong image earlier. The `mxxm/mx-llama.cpp:gfx906` image works fine.
